OADP-4655 Added section for kopia env variables (#80869)
openshift/openshift-docs: modules/oadp-kopia-algorithms-benchmarking.adoc

// Module included in the following assemblies:
//
// * backup_and_restore/application_backup_and_restore/installing/overriding-kopia-algorithms.adoc          

:_mod-docs-content-type: PROCEDURE
[id="oadp-kopia-algorithms-benchmarking_{context}"]
= Benchmarking Kopia hashing, encryption, and splitter algorithms

> You can run Kopia commands to benchmark the hashing, encryption, and splitter algorithms. Based on the benchmarking results, you can select the best algorithm suitable for your workload.

Reviewer: ```suggestion
You can run Kopia commands to benchmark the hashing, encryption, and splitter algorithms. Based on the benchmarking results, it is possible to select the most suitable algorithm for your workload.
```

> You can run Kopia commands to benchmark the hashing, encryption, and splitter algorithms. Based on the benchmarking results, it is possible to select the most suitable algorithm for your workload.

Reviewer 2: Should we mention that the benchmark needs to be ran from the cluster itself?

Reviewer 2: @weshayutin what is your take on it? Should we mention must-gather and the `kopia` CLI in there, so similarly to what was used in one of the tests:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: oadp-must-gather-pod
  labels:
    purpose: kopia-benchmark
spec:
  containers:
  - name: oadp-must-gather-container
    image: registry.redhat.io/oadp/oadp-mustgather-rhel8:v1.1
    command: ["sleep"]
    args: ["infinity"] 
```

Reviewer 3: @mpryc hold up.. you are saying there is a difference when one run benchmark from laptop vs. a pod?  This is news.   Help me understand that bit please.

Reviewer 2: **YES** The results  can vary significantly depending on the machine. Things like CPU speed, available RAM, disk speed, current I/O load and so on.

Here is an upstream comment from which hopefully makes things clear:
   https://kopia.discourse.group/t/questions-i-cannot-find-an-answer-to/428/6

With the above I think that benchmarks may and will vary depending on the resources, however the % increase or decrease between algorithms should be similar.

Author: @weshayutin @mpryc - I will add this information. Something like a note saying -  results differ from pod vs laptop and if you want to benchmark from a pod here is how you configure a pod - add the 'must-gather' config snippet . Thanks.

Author: Done.

.Prerequisites

* You have installed the {oadp-short} Operator.

> * You have installed the Kopia client.

Reviewer 3: for line #14 we *could* also note the kopia client is available on the must-gather image and can be setup as Michal demonstrated in the above comment.

```
apiVersion: v1
kind: Pod
metadata:
  name: oadp-mustgather-pod
  labels:
    purpose: user-interaction
spec:
  containers:
  - name: oadp-mustgather-container
    image: registry.redhat.io/oadp/oadp-mustgather-rhel9:v1.4
    command: ["sleep"]
    args: ["infinity"]

```

Author: Done.

* You have an application with persistent volumes running in a separate namespace.
* You have run a backup of the application with CSI snapshots.

.Procedure

> . Connect to the Kopia repository by running the following command:

Reviewer 2: This is an example, of course this depends on a specific s3 storage options, entire kopia CLI subset of commands. However I am not sure we want to point to that, we should mention that this is an example ? E.g. disable-tls may not be used every single time?:

https://kopia.io/docs/reference/command-line/common/repository-connect-s3/

// Module included in the following assemblies:
//
// * backup_and_restore/application_backup_and_restore/installing/overriding-kopia-algorithms.adoc          

:_mod-docs-content-type: PROCEDURE
[id="oadp-kopia-algorithms-benchmarking_{context}"]
= Benchmarking Kopia hashing, encryption, and splitter algorithms

> You can run Kopia commands to benchmark the hashing, encryption, and splitter algorithms. Based on the benchmarking results, it is possible to select the most suitable algorithm for your workload. In this procedure, you run the Kopia benchmarking commands from a pod on the cluster. The benchmarking results can vary depending on CPU speed, available RAM, disk speed, current I/O load, and so on.

Reviewer 4: ```suggestion
...it is possible to select the most suitable algorithm for your workload.
```

Does "it" refer to Kopia commands or the algorithms?

Author: Have reworded it to 'you can select..'.

.Prerequisites

> * You have installed the {oadp-short} Operator.

Reviewer 4: Should OADP be spelled out in it's full term on first instance on the document?

* You have an application with persistent volumes running in a separate namespace.

> * You have run a backup of the application with CSI snapshots.

Reviewer 4: Spell out "CSI" on fist instance.

Author: Done.

.Procedure

. Configure a pod as shown in the following example. Make sure you are using the `oadp-mustgather` image for {oadp-short} version 1.3 and later.
+
.Example pod configuration
+
[source,yaml]
----
apiVersion: v1
kind: Pod
metadata:
  name: oadp-mustgather-pod
  labels:
    purpose: user-interaction
spec:
  containers:
  - name: oadp-mustgather-container
    image: registry.redhat.io/oadp/oadp-mustgather-rhel9:v1.3
    command: ["sleep"]

>     args: ["infinity"]

Reviewer 4: ```suggestion
    args: ["infinity"]
# ...
```

Assuming the YAML is not complete in this example?

Author: The yaml is fully complete. I have myself tested the yaml on a QE cluster. Therefore, not including the '...'. Thanks.

----
<1> Specify the name of the YAML file for the pod configuration.
+
. Verify that the Security Context Constraints (SCC) on the pod is `anyuid`, so that Kopia can connect to the repository.
+ 
[source,terminal]
----
$ oc describe pod/oadp-mustgather-pod | grep scc
----
+
.Example output
+
[source,terminal]
----
openshift.io/scc: anyuid
----
+
. Connect to the pod via SSH by running the following command:
+ 
[source,terminal]
----
$ oc -n openshift-adp rsh pod/oadp-mustgather-pod
----
+
. Connect to the Kopia repository by running the following command:
+
[source,terminal]
----
sh-5.1# kopia repository connect s3 \
  --bucket=<bucket_name> \ # <1>
  --prefix=velero/kopia/<application_namespace> \ # <2>
  --password=static-passw0rd \ # <3>
  --access-key="<access_key>" \ # <4>
  --secret-access-key="<secret_access_key>" \ # <5>
  --endpoint=<bucket_endpoint> \ # <6>
----
<1> Specify the object storage provider bucket name.
<2> Specify the namespace for the application.
<3> This is the Kopia password to connect to the repository.
<4> Specify the object storage provider access key.
<5> Specify the object storage provider secret access key.

> <6> Specify the bucket endpoint. You do not need to specify the bucket endpoint, if you are using AWS S3 as the storage provider.

Reviewer 4: An attribute exists for AWS in it's short form. OADP docs also reference the `common-attributes.adoc` file?

![Screenshot from 2024-09-02 16-30-01](https://github.com/user-attachments/assets/003abf40-b8ee-4206-91a1-2922b0898b0b)

Author: done.

DYNAMIC                     747.6 MB/s count:107 min:9467 10th:2277562 25th:2971794 50th:4747177 75th:7603998 90th:8388608 max:8388608
DYNAMIC-128K-BUZHASH        718.5 MB/s count:3183 min:3076 10th:80896 25th:104312 50th:157621 75th:249115 90th:262144 max:262144
DYNAMIC-128K-RABINKARP      164.4 MB/s count:3160 min:9667 10th:80098 25th:106626 50th:162269 75th:250655 90th:262144 max:262144
# ...
FIXED-512K                  102.9 TB/s count:1024 min:524288 10th:524288 25th:524288 50th:524288 75th:524288 90th:524288 max:524288
FIXED-8M                    566.3 TB/s count:64 min:8388608 10th:8388608 25th:8388608 50th:8388608 75th:8388608 90th:8388608 max:8388608
-----------------------------------------------------------------
  0. FIXED-8M                  566.3 TB/s   count:64 min:8388608 10th:8388608 25th:8388608 50th:8388608 75th:8388608 90th:8388608 max:8388608
  1. FIXED-4M                  425.8 TB/s   count:128 min:4194304 10th:4194304 25th:4194304 50th:4194304 75th:4194304 90th:4194304 max:4194304
  # ...
 22. DYNAMIC-128K-RABINKARP    164.4 MB/s   count:3160 min:9667 10th:80098 25th:106626 50th:162269 75th:250655 90th:262144 max:262144

> ----

Reviewer 4: Maybe end the file with a blank line? Sometimes the customer portal can be funny with files with no empty space that impact the heading from rendering in the next file.
